File: androld/frist/app/src/main/java/z3d/res/BaseRes.java
package z3d.res;

import android.print.PrinterId;
import android.util.Log;


import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;
import java.util.zip.Inflater;
import java.util.zip.ZipFile;

import z3d.base.CallBackFun;

import z3d.base.ObjDataManager;
import z3d.engine.ResCount;
import z3d.base.ByteArray;
import z3d.filemodel.ParticleManager;
import z3d.material.MaterialManager;
import z3d.units.TimeUtil;

public class BaseRes extends ResCount {
    public static int IMG_TYPE = 1;
    public static int OBJS_TYPE = 2;
    public static int MATERIAL_TYPE = 3;
    public static int PARTICLE_TYPE = 4;
    public static int SCENE_TYPE = 5;
    public static int ZIP_OBJS_TYPE = 6;
    public static int PREFAB_TYPE = 1;
    public static int SCENE_PARTICLE_TYPE = 11;
    protected ByteArray _byte;
    protected int imgNum;
    protected int imgLoadNum;
    private CallBackFun imgFun;
    public int version;


    public void read( ){

        this.read(null);
    }
    public void read(CallBackFun bfun){

        this.imgFun=bfun;
        int fileType   = this._byte.readInt();
        Log.d("fileType",   fileType+"" );
        if (fileType == BaseRes.IMG_TYPE) {
            this.readImg();

        } else if (fileType == BaseRes.OBJS_TYPE) {
            // this.readObj(this._byte);
        } else if (fileType == BaseRes.MATERIAL_TYPE) {
            this.readMaterial();
        } else if (fileType == BaseRes.PARTICLE_TYPE) {
            this.readParticle();
        } else if (fileType == BaseRes.ZIP_OBJS_TYPE) {
            this.readZipObj();
        }
        if(this.imgFun!=null){
            this.imgFun.StateChange(true);
        }
    }
    private  void readMaterial()
    {
        int objNum=this._byte.readInt();
        int time = TimeUtil.getTimer();
        for (int i = 0; i < objNum; i++) {

            String url=this._byte.readUTF();
            int materialSize=this._byte.readInt();
            byte[] materialByte= this._byte.readBytes(materialSize);
            Log.d("material地址 ->",   url+" " );
            Log.d("material大小 ->",   materialSize+" " );

            MaterialManager.getInstance().addResByte(url, new ByteArray(materialByte));

         }

    }

    private  void readParticle()
    {
        int particleNum=this._byte.readInt();
        int time = TimeUtil.getTimer();
        for (int i = 0; i < particleNum; i++) {

            String url=this._byte.readUTF();
            int particSize=this._byte.readInt();
            byte[] particByte= this._byte.readBytes(particSize);
            Log.d("partic地址 ->",   url+" " );
            Log.d("partic大小 ->",   particSize+" " );

            ParticleManager.getInstance().addResByte(url, new ByteArray(particByte));

        }

    }

    /*
    public readParticle(): void {
        var objNum: number = this._byte.readInt();

        var time: number = TimeUtil.getTimer();

        for (var i: number = 0; i < objNum; i++) {
            var url: string = Scene_data.fileRoot + this._byte.readUTF();
            var size: number = this._byte.readInt();


            var dataByte: Pan3dByteArray = new Pan3dByteArray;
            dataByte.length = size;
            this._byte.readBytes(dataByte, 0, size)
            ParticleManager.getInstance().addResByte(url, dataByte);


        }


    }
    */
    private  void   readZipObj(){
        int zipLen = this._byte.readInt();
        byte[] zipByte= this._byte.readBytes(zipLen);
        byte[] outByte=this.getZipData(zipByte);
        this.readObj(new ByteArray(outByte));
    }
    public byte[] getZipData(byte[] sourcesByte) {
        int dataLength = 1024*1024;
        byte[] result=new byte[dataLength];
        try {
            Inflater decompresser=new Inflater();
            decompresser.setInput(sourcesByte,0,sourcesByte.length);
            int resultLength = decompresser.inflate(result);
            decompresser.end();
        } catch(Exception ex) {}
        return result;

    }

    private  void  readObj(ByteArray srcByte){
        int objNum  = srcByte.readInt();
        for (int i   = 0; i < objNum; i++) {
            int pos=srcByte.byteBuffer.position;
            String url = srcByte.readUTF();
            int objLen = srcByte.readInt();
            byte[] objByte= srcByte.readBytes(objLen);
            Log.d("obj地址 ->",   url+" " );
            Log.d("obj大小 ->",   objLen+" " );


            ObjDataManager.getInstance().loadObjCom(  objByte, url);


        }



    }

    public static void readIntForTwoByte(ByteArray  $byte, List indexs)
    {
        int iLen   = $byte.readInt();
        for (int i = 0; i < iLen; i++) {
            indexs.add($byte.readShort());
        }
    }

    public static List<Float>  readBytes2ArrayBuffer(ByteArray $byte, ByteBuffer $data,int dataWidth, int offset, int stride, int readType)
    {
        int verLength = $byte.readInt();

        List<Float>  list=new ArrayList();

        if (verLength <= 0) {
            return list;
        }

        float scaleNum=0;
        if (readType == 0) {
            scaleNum = $byte.readFloat();
        }
        float tempNum=0;
        int readNum   = verLength / dataWidth;

        for (int i = 0; i < readNum; i++) {
            int pos = stride * i + offset;
            for (int j = 0; j < dataWidth; j++) {
                if (readType == 0) {
                    tempNum=  $byte.readFloatTwoByte(scaleNum);
                    // $data.setFloat32((pos + j) * 4, $byte.readFloatTwoByte(scaleNum), true);
                } else if (readType == 1) {
                    // $data.setFloat32((pos + j) * 4, $byte.readFloatOneByte(), true);
                    tempNum=  $byte.readFloatOneByte();
                } else if (readType == 2) {
                    // $data.setFloat32((pos + j) * 4, $byte.readByte(), true);
                } else if (readType == 3) {
                    // $data.setFloat32((pos + j) * 4, ($byte.readByte() + 128) / 255, true);
                } else if (readType == 4) {
                    // $data.setFloat32((pos + j) * 4, $byte.readFloat(), true);
                }

                list.add(tempNum);

            }


        }
        return list;


    }
        public  void readImg()
    {
        this.imgNum = this._byte.readInt();
        this._byte.tracePostion("图片数量"+imgNum);
        this.imgLoadNum = 0;
        for (int i = 0; i < this.imgNum; i++) {
            String url   =  this._byte.readUTF();
            this._byte.tracePostion("b");
            int imgSize = this._byte.readInt();
            Log.d("图片地址 ->",   url+" " );
            Log.d("图片大小 ->",   imgSize+" " );
            if(imgSize>0){
                this._byte.readBytes(imgSize);

            }
        }
        Log.d("url ->",     "-----" );
    }



    /*
    public read($imgFun: Function = null): void {
            this._imgFun = $imgFun;
            var fileType: number = this._byte.readInt();
            if (fileType == BaseRes.IMG_TYPE) {
                if (Scene_data.supportBlob) {
                    this.readImg();
                } else {
                    this.readImgLow();
                }
            } else if (fileType == BaseRes.OBJS_TYPE) {
                this.readObj(this._byte);
            } else if (fileType == BaseRes.MATERIAL_TYPE) {
                this.readMaterial();
            } else if (fileType == BaseRes.PARTICLE_TYPE) {
                this.readParticle();
            } else if (fileType == BaseRes.ZIP_OBJS_TYPE) {
                this.readZipObj();
            }

        }
     */
}

File: h5gl/new/src/pan3d/res/BaseRes.ts
module Pan3d {
    export class BaseRes extends ResCount {


        public static IMG_TYPE: number = 1;
        public static OBJS_TYPE: number = 2;
        public static MATERIAL_TYPE: number = 3;
        public static PARTICLE_TYPE: number = 4;
        public static SCENE_TYPE: number = 5;
        public static ZIP_OBJS_TYPE: number = 6;
        public static PREFAB_TYPE: number = 1;
        public static SCENE_PARTICLE_TYPE: number = 11;


        protected _byte: Pan3dByteArray
        protected imgNum: number;
        protected imgLoadNum: number;
        protected _imgComplete: boolean;

        //public imgAry: Array<string>;
        //public objAry: Array<string>;
        //public materialAry: Array<string>;
        //public particleAry: Array<string>;
        protected _progressFun: Function;
        //public useNum: number;
        public version: number;
        public _imgFun: Function;

        private allImgBytes: number = 10000000;

        constructor(value:Scene3D) {
            super(value);
        }

        public read($imgFun: Function = null): void {
            this._imgFun = $imgFun;
            var fileType: number = this._byte.readInt();
            if (fileType == BaseRes.IMG_TYPE) {
                if (this.scene3D.supportBlob) {
                    this.readImg();
                } else {
                    // this.readImgLow();
                }
            } else if (fileType == BaseRes.OBJS_TYPE) {
                this.readObj(this._byte);
            } else if (fileType == BaseRes.MATERIAL_TYPE) {
                this.readMaterial();
            } else if (fileType == BaseRes.PARTICLE_TYPE) {
                this.readParticle();
            } else if (fileType == BaseRes.ZIP_OBJS_TYPE) {
                this.readZipObj();
            }

        }

        public readZipObj(): void {
            var zipLen: number = this._byte.readInt();
            var aryBuf: ArrayBuffer = this._byte.buffer.slice(this._byte.position, this._byte.position + zipLen);
            this._byte.position += zipLen;
            var zipedBuf: ArrayBuffer = unZip(aryBuf)
            var newByte: Pan3dByteArray = new Pan3dByteArray(zipedBuf);
            this.readObj(newByte);
        }

        public readImg(): void {
            this.imgNum = this._byte.readInt();
            this.imgLoadNum = 0;
            //this.imgAry = new Array;


            // var time: number = TimeUtil.getTimer();

            var ary: Array<Blob> = new Array;
            var urlAry: Array<string> = new Array;

            for (var i: number = 0; i < this.imgNum; i++) {
                var url: string = this.scene3D.fileRoot + this._byte.readUTF();
                var imgSize: number = this._byte.readInt();
                
            

                var imgAryBuffer: ArrayBuffer = this._byte.buffer.slice(this._byte.position, this._byte.position + imgSize);
                this._byte.position += imgSize;


                var blob: Blob = new Blob([imgAryBuffer], { type: "application/octet-binary" });
                ary.push(blob);
                urlAry.push(url);

            }

            for (var i: number = 0; i < ary.length; i++) {
                var img: any = new Image();
                img.url = urlAry[i];
                img.onload = (evt: Event) => {
                    this.loadImg(evt.target);
                    var etimg: any = evt.target;
                    URL.revokeObjectURL(etimg.src);
                }
                img.src = URL.createObjectURL(ary[i]);
            }

        }

  
        public loadImg(img: any): void {
            // TextureManager.getInstance().addRes(img.url, img);

            this.countImg();
        }

        public addImg($url: string, img: any): void {
            // TextureManager.getInstance().addRes($url, img);

            this.countImg();
        }

        public countImg(): void {
            this.imgLoadNum++;
            if (this.imgLoadNum == this.imgNum) {
                this._imgComplete = true;
                this.allResCom();
            }
        }

        public readObj($srcByte: Pan3dByteArray): void {
            var objNum: number = $srcByte.readInt();

            for (var i: number = 0; i < objNum; i++) {
                var url: string = this.scene3D.fileRoot + $srcByte.readUTF();
                var size: number = $srcByte.readInt();
                var newByte: Pan3dByteArray = new Pan3dByteArray();
                newByte.length = size;
                $srcByte.readBytes(newByte, 0, size);
               var objData: ObjData = this.scene3D.objDataManager.loadObjCom(newByte.buffer, url);
            }

            if (this._imgFun) {
                this._imgFun();
            }

        }
        public readMaterial(): void {
            var objNum: number = this._byte.readInt();
            //this.materialAry = new Array;

            // var time: number = TimeUtil.getTimer();

            for (var i: number = 0; i < objNum; i++) {
                var url: string = this.scene3D.fileRoot + this._byte.readUTF();
                var size: number = this._byte.readInt();

                var dataByte: Pan3dByteArray = new Pan3dByteArray;
                dataByte.length = size;
                this._byte.readBytes(dataByte, 0, size)
                // MaterialManager.getInstance().addResByte(url, dataByte);
          

            }
      
        }

        public readParticle(): void {
            var objNum: number = this._byte.readInt();
          

            for (var i: number = 0; i < objNum; i++) {
                var url: string = this.scene3D.fileRoot + this._byte.readUTF();
                var size: number = this._byte.readInt();


                var dataByte: Pan3dByteArray = new Pan3dByteArray;
                dataByte.length = size;
                this._byte.readBytes(dataByte, 0, size)
                // ParticleManager.getInstance().addResByte(url, dataByte);

 
            }

             
        }
        //读材质参数
        public readMaterialInfo(): Array<any> {

            var len: number = this._byte.readInt();
            if (len > 0) {
                var $arr: Array<any> = new Array
                for (var i: number = 0; i < len; i++) {
                    var $temp: any = new Object();
                    $temp.type = this._byte.readInt()
                    $temp.name = this._byte.readUTF()
                    if ($temp.type == 0) {
                        $temp.url = this._byte.readUTF()
                    }
                    if ($temp.type == 1) {
                        $temp.x = this._byte.readFloat()
                    }
                    if ($temp.type == 2) {
                        $temp.x = this._byte.readFloat()
                        $temp.y = this._byte.readFloat()
                    }
                    if ($temp.type == 3) {
                        $temp.x = this._byte.readFloat()
                        $temp.y = this._byte.readFloat()
                        $temp.z = this._byte.readFloat()
                    }

                    $arr.push($temp)
                }
                return $arr
            } else {
                return null
            }
        }
        //读取浮点数据，两个字节
        public static readFloatTwoByte(byte: Pan3dByteArray, vertices: Array<number>): void {


            var verLength: number = byte.readInt();
            if (verLength > 0) {
                var $scaleNum: number = byte.readFloat();
                vertices.length = 0
                for (var i: number = 0; i < verLength; i++) {
                    vertices.push(byte.readFloatTwoByte($scaleNum))

                }
            }

        }

        //读取一个字节的LightMap
        public static readFloatOneByte(byte: Pan3dByteArray, vertices: Array<number>): void {
            var verLength: number = byte.readInt();
            if (verLength > 0) {
                for (var i: number = 0; i < verLength; i++) {
                    vertices.push((byte.readByte() + 128) / 256)
                }
            }

        }

        public static readIntForTwoByte(byte: Pan3dByteArray, indexs: Array<number>): void {
            var iLen: number = byte.readInt();
            for (var i: number = 0; i < iLen; i++) {
                indexs.push(byte.readShort());
            }
        }

        public static readIntForOneByte(byte: Pan3dByteArray, indexs: Array<number>): void {
            var iLen: number = byte.readInt();
            for (var i: number = 0; i < iLen; i++) {
                indexs.push(byte.readByte());
            }
        }


        /**
         * $readType
         * 0 readFloatTwoByte
         * 1 readFloatOneByte
         * 2 readIntForOneByte
         *  */
        public static readBytes2ArrayBuffer($byte: Pan3dByteArray, $data: DataView, $dataWidth: number, $offset: number, $stride: number, $readType: number = 0): void {
            var verLength: number = $byte.readInt();

            if (verLength <= 0) {
                return;
            }

            var scaleNum: number;
            if ($readType == 0) {
                scaleNum = $byte.readFloat();
            }

            var readNum: number = verLength / $dataWidth;
            for (var i: number = 0; i < readNum; i++) {
                var pos: number = $stride * i + $offset;
                for (var j: number = 0; j < $dataWidth; j++) {

                    if ($readType == 0) {
                        $data.setFloat32((pos + j) * 4, $byte.readFloatTwoByte(scaleNum), true);
                    } else if ($readType == 1) {
                        $data.setFloat32((pos + j) * 4, $byte.readFloatOneByte(), true);
                    } else if ($readType == 2) {
                        $data.setFloat32((pos + j) * 4, $byte.readByte(), true);
                    } else if ($readType == 3) {
                        $data.setFloat32((pos + j) * 4, ($byte.readByte() + 128) / 255, true);
                    } else if ($readType == 4) {
                        $data.setFloat32((pos + j) * 4, $byte.readFloat(), true);
                    }


                }

            }



        }



        //读取材质参数
        public static readMaterialParamData(byte: Pan3dByteArray): Array<any> {
            var mpNum: number = byte.readInt();
            if (mpNum > 0) {
                var mpAry: Array<any> = new Array;
                for (var j: number = 0; j < mpNum; j++) {
                    var obj: any = new Object;
                    obj.name = byte.readUTF();
                    obj.type = byte.readByte();
                    if (obj.type == 0) {
                        obj.url = byte.readUTF();
                    } else if (obj.type == 1) {
                        obj.x = byte.readFloat();
                    } else if (obj.type == 2) {
                        obj.x = byte.readFloat();
                        obj.y = byte.readFloat();
                    } else if (obj.type == 3) {
                        obj.x = byte.readFloat();
                        obj.y = byte.readFloat();
                        obj.z = byte.readFloat();
                    }
                    mpAry.push(obj);
                }
                return mpAry
            }
            return null
        }
        public allResCom(): void {
            if (this._imgFun) {
                this._imgFun();
            }
        }
    }
}